src/Data/Nat/Primality.agda

```diff
@@ -8,80 +8,290 @@
+
+instance
+  nonUnit[0] : NonUnit 0
+  nonUnit[0] = _
+
+  nonTrivial⇒nonUnit : .{{NonTrivial n}} → NonUnit n
+  nonTrivial⇒nonUnit {n = 2+ _} = _
+
+nonUnit⇒≢1 : .{{NonUnit n}} → n ≢ 1
+nonUnit⇒≢1 ⦃()⦄ refl
+instance
+  nonTrivial : NonTrivial (2+ n)
+  nonTrivial = _
+
+nonTrivial⇒≢1 : .{{NonTrivial n}} → n ≢ 1
+nonTrivial⇒≢1 ⦃()⦄ refl
+
+nonTrivial⇒nonZero : .{{NonTrivial n}} → NonZero n
+nonTrivial⇒nonZero {n = 2+ k} = _
+
+pattern 1<2+n {n} = s<s (z<s {n})
+
+nonTrivial⇒n>1 : .{{NonTrivial n}} → 1 < n
+nonTrivial⇒n>1 {n = 2+ _} = 1<2+n
+
+n>1⇒nonTrivial : 1 < n → NonTrivial n
+n>1⇒nonTrivial 1<2+n = _
+
 
 ------------------------------------------------------------------------
 -- Definitions
 
+-- Definition of having a non-trivial divisor below a given bound
+
+record BoundedComposite (k n d : ℕ) : Set where
+  constructor boundedComposite
+  field
+    {{nt}} : NonTrivial d
+    d<k : d < k
+    d∣n : d ∣ n
+
+-- smart constructors
+
+boundedComposite≢ : .{{NonZero n}} → {{NonTrivial d}} →
+                    d ≢ n → d ∣ n → BoundedComposite n n d
+boundedComposite≢ d≢n d∣n = boundedComposite (≤∧≢⇒< (∣⇒≤ d∣n) d≢n) d∣n
+
+boundedComposite>1 : 1 < d → d < n → d ∣ n → BoundedComposite n n d
+boundedComposite>1 1<d = boundedComposite ⦃ n>1⇒nonTrivial 1<d ⦄
+
 -- Definition of compositeness
 
-Composite : ℕ → Set
-Composite 0 = ⊥
-Composite 1 = ⊥
-Composite n = ∃ λ d → 2 ≤ d × d < n × d ∣ n
+Composite : Pred ℕ _
+Composite n = ∃⟨ BoundedComposite n n ⟩
+
+-- smart constructor
+
+composite : .{{NonZero n}} → {{NonTrivial d}} →
+            d ≢ n → d ∣ n → Composite n
+composite {d = d} d≢n d∣n = d , boundedComposite≢ d≢n d∣n
+
+-- Definition of 'rough': a number is k-rough
+-- if all its non-trivial factors d are bounded below by k
+
+Rough : ℕ → Pred ℕ _
+Rough k n = ∀[  ¬_ ∘ BoundedComposite k n ]
+
+-- Definition of primality: complement of Composite
+-- Constructor `prime` takes a proof isPrime that
+-- NonTrivial p is p-Rough, and thereby enforces that
+-- p is a fortiori NonZero and NonUnit
+
+record Prime (p : ℕ) : Set where
+  constructor prime
+  field
+    {{nt}}  : NonTrivial p
+    isPrime : Rough p p
+
+-- Definition of irreducibility
+
+Irreducible : Pred ℕ _
+Irreducible n = ∀[ irreducible n ]
+  where
+  irreducible : ℕ → Pred ℕ _
+  irreducible n d = d ∣ n → d ≡ 1 ⊎ d ≡ n
+
+
+------------------------------------------------------------------------
+-- Basic properties of Rough
+
+-- 1 is always rough
+rough-1 : ∀ k → Rough k 1
+rough-1 _ {2+ _} (boundedComposite _ d∣1@(divides q@(suc _) ()))
+
+-- Any number is 0-, 1- and 2-rough,
+-- because no non-trivial factor d can be less than 0, 1, or 2
+0-rough : Rough 0 n
+0-rough (boundedComposite () _)
+
+1-rough : Rough 1 n
+1-rough (boundedComposite ⦃()⦄ z<s _)
+
+2-rough : Rough 2 n
+2-rough (boundedComposite ⦃()⦄ (s<s z<s) _)
 
--- Definition of primality.
+-- If a number n > 1 is k-rough, then k ≤ n
+rough⇒≤ : ⦃ NonTrivial n ⦄ → Rough k n → k ≤ n
```

Comment:
First argument should be irrelevant?

Reply:
So... the (anonymous, but relevant) instance *is* passed from the input to the constructor in the output, so it appears to be 'relevant', and indeed the constructor turns yellow if the input is marked irrelevant... But your suggestion about the divisor `d` being a record field might change that.

Reply:
No, all it required was to make the instance field in the record irrelevant...